python_scripts/generate_test_data.py

```diff
@@ -0,0 +1,458 @@
+
+        # Get a cursor
+        cursor = db_conn.cursor(dictionary=True, buffered=True)
+
+        # Call stored procedure with source barcode
+        args = [barcode]
+        print("barcode for proc = %s" % barcode)
+        cursor.callproc('getPickableSamplesForSourcePlate', args)
+
+        # Process the result into a dictionary
+        for result in cursor.stored_results(): # only one result
+            wells = result.fetchall() # array of key value pairs
+            for well in wells:
+                pickable_samples.append(
+                    {
+                        'source_coordinate': well['coordinate'],
+                        'sample_id': well['sample_id'],
+                        'rna_id': well['rna_id'],
+                        'lab_id': well['lab_id']
+                    }
+                )
+
+    except Exception as e:
+        print(e)
+    else:
+        print("Fetched pickable samples for source plate %s" % barcode)
+    finally:
+        cursor.close()
+        db_conn.close()
+        return pickable_samples
+
+# Update the destination plate well for a picked source plate well
+def update_destination_plate_well_for_source(system_name, system_run_id, destination_barcode, destination_coordinate, source_barcode, source_coordinate):
+    try:
+        # open DB connection
+        db_conn = create_database_connection()
+
+        # Get a cursor
+        cursor = db_conn.cursor()
+
+        # Call stored procedure with parameters
+        args = [system_name, system_run_id, destination_barcode, destination_coordinate, source_barcode, source_coordinate]
+        cursor.callproc('updateDestinationPlateWellWithSource', args)
+
+        # commit changes
+        db_conn.commit()
+    except Exception as e:
+        print(e)
+    else:
+        print("Updated destination well %s with picked source %s coordinate %s" % (destination_coordinate, source_barcode, source_coordinate))
+    finally:
+        cursor.close()
+        db_conn.close()
+
+# Update the run record at the end of the run
+def update_run_record(system_name, system_run_id, new_state):
+    try:
+        # open DB connection
+        db_conn = create_database_connection()
+
+        # Get a cursor
+        cursor = db_conn.cursor()
+
+        # Call stored procedure with parameters
+        args = [system_name, system_run_id, new_state]
+        cursor.callproc('updateRunState', args)
+
+        # commit changes
+        db_conn.commit()
+    except Exception as e:
+        print(e)
+    else:
+        print("Updated run record id %s with state %s" % (system_run_id, new_state))
+    finally:
+        cursor.close()
+        db_conn.close()
+
+# Fetch the last run id used
+def get_last_run_id() -> int:
+    last_run_id = 0
+    try:
+        # open DB connection
+        db_conn = create_database_connection()
+
+        # Get a cursor
+        cursor = db_conn.cursor()
+
+        sql = "SELECT MAX(system_run_id) FROM %s.automation_system_runs;" % DB_NAME
```

Comment:
This could be moved to stored procedure too

Reply:
Wasn't sure Biosero would need this one as a stored proc, but could do.